File: app/src/main/java/com/timetracker/dao/CategoryDao.java
package com.timetracker.dao;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

import com.timetracker.db.CategoriesContract;
import com.timetracker.entities.Action;
import com.timetracker.entities.Category;

import java.util.ArrayList;
import java.util.Date;
import java.util.HashSet;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Stream;

import static com.timetracker.db.CategoriesContract.CategoryEntry.*;

public class CategoryDao {

    private final SQLiteOpenHelper dbHelper;

    public CategoryDao(SQLiteOpenHelper dbHelper) {
        this.dbHelper = dbHelper;
    }

    public Stream<Category> list() {
        List<Category> result = new ArrayList<>();
        try(Cursor cursor = listCursor()) {
            while(cursor.moveToNext()) {
                result.add(
                        new Category(
                                cursor.getInt(cursor.getColumnIndex(_ID)),
                                cursor.getString(cursor.getColumnIndex(COLUMN_NAME_NAME))
                        )
                );
            }
        }
        return result.stream();
    }

    public Cursor listCursor() {
        return dbHelper
            .getReadableDatabase()
            .rawQuery("select * from " + TABLE_NAME, null);
    }

    public void save(Category.CreateCategory createCategory) {
        try (SQLiteDatabase db = dbHelper.getWritableDatabase()) {
            ContentValues contentValues = new ContentValues();
            contentValues.put(CategoriesContract.CategoryEntry.COLUMN_NAME_NAME, createCategory.name);
            db.insert(CategoriesContract.CategoryEntry.TABLE_NAME, null, contentValues);
        }
    }

    public Cursor findByIdCursor(Integer categoryId) {
        return dbHelper.getReadableDatabase()
                .rawQuery(String.format("select * from %s where %s=%s",
                        TABLE_NAME, _ID, categoryId), null);
    }

    public Optional<Category> findById(Integer categoryId) {
        Cursor cursor = findByIdCursor(categoryId);
        if (cursor.moveToNext()) return Optional.of(currentCursorStateToCategory(cursor));
        else return Optional.empty();
    }

    private Category currentCursorStateToCategory(Cursor cursor) {
        return new Category(
                cursor.getInt(cursor.getColumnIndex(_ID)),
                cursor.getString(cursor.getColumnIndex(COLUMN_NAME_NAME))
        );
    }

}

File: app/src/main/java/com/timetracker/dao/CategoryDao.kt
package com.timetracker.dao

import android.content.ContentValues
import android.database.Cursor
import android.database.sqlite.SQLiteOpenHelper
import android.provider.BaseColumns

import com.timetracker.entities.Category

import com.timetracker.db.entries.CategoryEntry as CE
import java.util.*

class CategoryDao(private val dbHelper: SQLiteOpenHelper) {

    fun list(): List<Category> {
        val result = ArrayList<Category>()
        listCursor().use { cursor ->
            while (cursor.moveToNext()) {
                result.add(
                        Category(
                                cursor.getInt(cursor.getColumnIndex(BaseColumns._ID)),
                                cursor.getString(cursor.getColumnIndex(CE.COLUMN_NAME_NAME))
                        )
                )
            }
        }
        return result
    }

    fun listCursor(): Cursor {
        return dbHelper
                .readableDatabase
                .rawQuery("select * from " + CE.TABLE_NAME, null)
    }

    fun save(createCategory: Category.CreateCategory) {
        dbHelper.writableDatabase.use { db ->
            val contentValues = ContentValues()
            contentValues.put(CE.COLUMN_NAME_NAME, createCategory.name)
            db.insert(CE.TABLE_NAME, null, contentValues)
        }
    }

    fun findByIdCursor(categoryId: Int?): Cursor {
        return dbHelper.readableDatabase
                .rawQuery(String.format("select * from %s where %s=%s",
                        CE.TABLE_NAME, BaseColumns._ID, categoryId), null)
    }

    fun findById(categoryId: Int?): Category? {
        val cursor = findByIdCursor(categoryId)
        if (cursor.moveToNext())
            return currentCursorStateToCategory(cursor)
        else
            return null
    }

    fun delete(categoryId: Int) {
        dbHelper.writableDatabase.use { db ->
            db.delete(CE.TABLE_NAME, "${BaseColumns._ID} = $categoryId", null)
        }
    }

    private fun currentCursorStateToCategory(cursor: Cursor): Category {
        return Category(
                cursor.getInt(cursor.getColumnIndex(BaseColumns._ID)),
                cursor.getString(cursor.getColumnIndex(CE.COLUMN_NAME_NAME))
        )
    }

}
